extensions/mongodb/src/main/java/com/hazelcast/jet/mongodb/WriteMongoP.java

```diff
@@ -0,0 +1,589 @@
+            String documentIdentityFieldName
+    ) {
+       this(clientSupplier, new FunctionalCollectionPicker<>(databaseNameSelectFn, collectionNameSelectFn),
+               documentType, documentIdentityFn, replaceOptionAdjuster, documentIdentityFieldName);
+    }
+
+    /**
+     * Creates a new processor that will always insert to the same database and collection.
+     */
+    public WriteMongoP(
+            SupplierEx<? extends MongoClient> clientSupplier,
+            CollectionPicker<I> picker,
+            Class<I> documentType,
+            FunctionEx<I, Object> documentIdentityFn,
+            ConsumerEx<ReplaceOptions> replaceOptionAdjuster,
+            String documentIdentityFieldName) {
+        this.connection = new MongoDbConnection(clientSupplier, client -> {
+        });
+        this.documentIdentityFn = documentIdentityFn;
+        this.documentIdentityFieldName = documentIdentityFieldName;
+        this.documentType = documentType;
+        this.collectionPicker = picker;
+
+        ReplaceOptions options = new ReplaceOptions().upsert(true);
+        if (replaceOptionAdjuster != null) {
+            replaceOptionAdjuster.accept(options);
+        }
+        this.replaceOptions = options;
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) {
+        logger = context.logger();
+
+        int processorIndex = context.globalProcessorIndex();
+        transactionUtility = new UnboundedTransactionsProcessorUtility<>(
+                getOutbox(),
+                context,
+                context.processingGuarantee(),
+                () -> new MongoTransactionId(context, processorIndex),
+                txId -> new MongoTransaction(txId, logger),
+                txId -> {
+                    MongoTransaction mongoTransaction = activeTransactions.get(txId);
+                    refreshTransaction(mongoTransaction, true);
+                },
+                () -> {
+                    for (MongoTransaction tx : activeTransactions.values()) {
+                        refreshTransaction(tx, false);
+                    }
+                }
+        );
+    }
+
+    @Override
+    public boolean tryProcessWatermark(@Nonnull Watermark watermark) {
+        return true;
+    }
+
+    @Override
+    protected boolean tryProcess(int ordinal, @Nonnull Object item) {
+        return transactionUtility.tryProcess();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void process(int ordinal, @Nonnull Inbox inbox) {
+        if (!connection.reconnectIfNecessary()) {
+            return;
+        }
+
+        try {
+            MongoTransaction mongoTransaction = transactionUtility.activeTransaction();
+
+            ArrayList<I> items = new ArrayList<>();
+            for (Object item : inbox) {
+                items.add((I) item);
+                if (items.size() >= MAX_BATCH_SIZE) {
+                    break;
+                }
+            }
```

Comment:
No items = zero  items and leaving early.

I'm not using drainTo, because in case of exception, we want to retry the transaction. If we drain the inbox, then you have to deal with given items in this invocation of "process", but you can also leave some items in the inbox and those will be checked the next time process is invoked.